ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Impl.hs

```diff
@@ -357,17 +367,21 @@ garbageCollectImpl env@VolatileDBEnv{..} slot =
 --
 -- This may throw an FsError.
 tryCollectFile :: forall m h blockId
-               .  (MonadThrow m, Ord blockId)
+               .  (IOLike m, Ord blockId)
                => HasFS m h
                -> VolatileDBEnv m blockId
                -> SlotNo
-               -> InternalState blockId h
-               -> (FileId, FileInfo blockId)
-               -> m (InternalState blockId h)
+               -> InternalState blockId m h
+               -> (FileId, FileInfo blockId h m)
+               -> m (InternalState blockId m h)
 tryCollectFile hasFS@HasFS{..} env slot st@InternalState{..} (fileId, fileInfo) =
     if  | not canGC     -> return st
         | not isCurrent -> do
-            removeFile $ filePath fileId
+            let file = filePath fileId
+            -- First close the read handle.
+            forM_ (Index.getHandle fileId _currentMap) releaseHandle
```

Comment:
I had written an extensive comment with tables and all about a race condition here, but GitHub lost it :cry:. A simpler rewrite:

There's a race condition here. Consider the following sequence of events:

1. Thread A calls `getBlock`, the state is read in `withState` using `readMVar`. Note that the state is still left in the `MVar`. The file to read the block from is file X. Before thread A can read the block using the handle for file X, the next thing happens:
2. Thread B calls `garbageCollect`. It obtains the same state from the `MVar` as thread A did. It decides to garbage collect file X, so it releases/closes its handle.
3. Thread A tries to read its block from the handle for file X that thread B just closed. Error: handle closed :boom:

Note that the closing a handle that is being used blocks until it is no longer used, but that doesn't help here, because we'll use the handle after it has been closed.

To be entirely safe  `getBlock` and `garbageCollect` should not happen concurrently. This could be enforced by something akin to `ReadWriteVar`, but we would need to roll our own version as we're using the IOSim.

Moreover, `getBlock` and `putBlock` *can* happen concurrently. `putBlock` and `garbageCollect` can *almost* happen concurrently, as it is very unlikely that we need to garbage collect the file we're appending to. Maybe it is even worth to make this impossible?

Reply:
This is similar to what I try to explain here https://github.com/input-output-hk/ouroboros-network/pull/1165#discussion_r345673029.
But I guess this situation is expected if we want concurrent reads: If the user tries to get a block, while another thread makes this block obsolete by garbage collection, this should be an error. I also tried to control this situation: by using exclusive locks in the fs layer, read(2) and close(2) can't happen concurrently, which is an undefined behavior.

Maybe it's worth improving the api in these case: for example wrapping this exception with a better `VolatileDBError` or just return Nothing. I like the latter a bit more, since getBlock returns Nothing when a block is not found, however it has the disadvantage that it may hide other `handle closed` errors for other reasons. In my opinion, after a `handle closed` error, the best solution is to read the InternalState MVar again,  and see if the file is really garbage collected and only in this case return Nothing.

> putBlock and garbageCollect can almost happen concurrently, as it is very unlikely that we need to garbage collect the file we're appending to

This shouldn't be an issue, since these operations are serialized by the `InternalState` MVar

Reply:
> This is similar to what I try to explain here [#1165 (comment)](https://github.com/input-output-hk/ouroboros-network/pull/1165#discussion_r345673029).
> But I guess this situation is expected if we want concurrent reads: If the user tries to get a block, while another thread makes this block obsolete by garbage collection, this should be an error. I also tried to control this situation: by using exclusive locks in the fs layer, read(2) and close(2) can't happen concurrently, which is an undefined behavior.

Should it be an error, though? Returning `Nothing`, as you say in the next paragraph, seems like a reasonable thing to do.

> > This could be enforced by something akin to ReadWriteVar, but we would need to roll our own version as we're using the IOSim.
> 
> I'm afraid adding another lock would complicate things a bit too much.

True.

> 
> Maybe it's worth improving the api in these case: for example wrapping this exception with a better `VolatileDBError` or just return Nothing. I like the latter a bit more, since getBlock returns Nothing when a block is not found, however it has the disadvantage that it may hide other `handle closed` errors for other reasons. In my opinion, after a `handle closed` error, the best solution is to read the InternalState MVar again, and see if the file is really garbage collected and only in this case return Nothing.

That sounds good to me. If you implement this, add a comment to `getBlockImpl` saying this should not loop because a handle is only closed when the file is garbage collected (so we will find `Nothing` in the recursive call) or when the VolatileDB itself is closed (so we will get a `ClosedDBError` in the recursive call). (Grep for `errLoad` to see an example of where I only catch a certain error.)

> > putBlock and garbageCollect can almost happen concurrently, as it is very unlikely that we need to garbage collect the file we're appending to
> 
> This shouldn't be an issue, since these operations are serialized by the `InternalState` MVar

Yes, but my point is actually the opposite: we could allow them to happen concurrently if we make it impossible to garbage-collect the current file. Then garbage collection wouldn't block adding new blocks, which would speed up syncing, as adding a block to the VolatileDB is a synchronous step in the syncing process.

However, I'm inclined to think garbage collection can happen so fast that we don't need to bother with this (added complexity). Maybe if we measure this in the future with some benchmarks and we see a GC causing a noticeable drop in sync speed, we can think about it again.

Nonetheless, I think the current file should be never be garbage collected. In all properly configured VolatileDBs, this should not never be needed anyway. Getting rid of this special case (+ `reOpenFile`) will simplify things. Do you agree?